rust/datafusion/src/execution/physical_plan/merge.rs

```diff
@@ -64,33 +75,51 @@ struct MergePartition {
     schema: SchemaRef,
     /// Input partitions
     partitions: Vec<Arc<dyn Partition>>,
+    /// Maximum number of concurrent threads
+    max_concurrency: usize,
+}
+
+fn collect_from_thread(
+    thread: JoinHandle<Result<Vec<RecordBatch>>>,
+    combined_results: &mut Vec<Arc<RecordBatch>>,
+) -> Result<()> {
+    match thread.join() {
+        Ok(join) => {
+            join?
+                .iter()
+                .for_each(|batch| combined_results.push(Arc::new(batch.clone())));
+            Ok(())
+        }
+        Err(e) => Err(ExecutionError::General(format!(
+            "Error collecting batches from thread: {:?}",
+            e
+        ))),
+    }
 }
 
 impl Partition for MergePartition {
     fn execute(&self) -> Result<Arc<Mutex<dyn RecordBatchReader + Send + Sync>>> {
-        let threads: Vec<JoinHandle<Result<Vec<RecordBatch>>>> = self
-            .partitions
-            .iter()
-            .map(|p| {
-                let p = p.clone();
-                thread::spawn(move || {
-                    let it = p.execute()?;
-                    common::collect(it)
-                })
-            })
-            .collect();
+        let mut combined_results: Vec<Arc<RecordBatch>> = vec![];
+        let mut threads: Vec<JoinHandle<Result<Vec<RecordBatch>>>> = vec![];
+
+        for partition in &self.partitions {
+            // limit number of concurrent threads
+            if threads.len() == self.max_concurrency {
```

Comment:
This approach (to wait for an arbitrary thread (thread 0) to be done) can lead to cases where the thread you are waiting on happens to have a large amount of work but other threads are waiting on the next partitions. Thus you can block here with only one thread working.

Another approach that would avoid that starvation would be to put the partitions into a `Arc<Mutex<Channel>>` and then have the threads each read and process partitions from that channel. 

I can try and whip that up if you would like a specific code example of what I am talking about

Reply:
Sure, I use that approach already in Ballista. I can make that change here as well.

Reply:
This is now implemented.